pkg/controllers/vdb/passwordsecret_reconciler.go

```diff
@@ -63,89 +72,221 @@ func (a *PasswordSecretReconciler) Reconcile(ctx context.Context, _ *ctrl.Reques
+			return ctrl.Result{}, nil
+		}
+	}
+
+	// If everything is up-to-date, no-op
+	outdatedSandboxes := a.getSandboxesNeedingUpdating()
+	if a.statusMatchesSpec(vapi.MainCluster) && len(outdatedSandboxes) == 0 {
 		return ctrl.Result{}, nil
 	}
 
+	// Perform password update (main cluster or sandbox)
 	res, err := a.updatePasswordSecret(ctx)
 	if verrors.IsReconcileAborted(res, err) {
 		return res, err
 	}
 
-	return ctrl.Result{}, a.updatePasswordSecretStatus(ctx)
+	// For main cluster, ensure status and trigger follow-ups
+	if sbName == vapi.MainCluster {
+		if err := a.updatePasswordSecretStatus(ctx); err != nil {
+			return ctrl.Result{}, err
+		}
+		// Always trigger any out-of-date sandboxes
+		return a.triggerOutOfDateSandboxes(ctx, outdatedSandboxes)
+	}
+
+	// Unset sandbox annotation
+	if _, ok := a.ConfigMap.Annotations[vmeta.SandboxControllerPasswordChangeTriggerID]; ok {
+		delete(a.ConfigMap.Annotations, vmeta.SandboxControllerPasswordChangeTriggerID)
+		if err := a.Rec.GetClient().Update(ctx, a.ConfigMap); err != nil {
+			a.Log.Error(err, "failed to remove password change trigger annotation")
+			return ctrl.Result{}, err
+		}
+		a.Log.Info("Removed password change trigger annotation from sandbox configmap")
+	}
+
+	return ctrl.Result{}, a.updatePasswordSecretStatusForSandbox(ctx, sbName)
 }
 
-// statusMatchesSpec checks if the password secret status is the same to spec or not
-func (a *PasswordSecretReconciler) statusMatchesSpec() bool {
-	return a.Vdb.Spec.PasswordSecret == *a.Vdb.Status.PasswordSecret
+// ensureStatusInitialized sets up initial status if not yet populated for main cluster or sandbox.
+func (a *PasswordSecretReconciler) ensureStatusInitialized(ctx context.Context, sbName string) error {
+	if sbName == vapi.MainCluster && a.Vdb.Status.PasswordSecret == nil {
+		return a.updatePasswordSecretStatus(ctx)
+	}
+	if sbName != vapi.MainCluster {
+		sandboxStatus := a.Vdb.GetSandboxStatus(sbName)
+		if sandboxStatus == nil {
+			return fmt.Errorf("could not find sandbox %q in status", sbName)
+		}
+		if sandboxStatus.PasswordSecret == nil {
+			return a.updatePasswordSecretStatusForSandbox(ctx, sbName)
+		}
+	}
+	return nil
 }
 
-// updatePasswordSecret will update the password secret in the database
+// statusMatchesSpec checks if the password secret status is the same as spec for either main cluster
+// or a specific sandbox.
+func (a *PasswordSecretReconciler) statusMatchesSpec(sbName string) bool {
+	if sbName == vapi.MainCluster {
+		if a.Vdb.Status.PasswordSecret == nil {
+			return false
+		}
+		return a.Vdb.Spec.PasswordSecret == *a.Vdb.Status.PasswordSecret
+	}
+
+	return a.Vdb.Spec.PasswordSecret == a.Vdb.GetPasswordSecretForSandbox(sbName)
+}
+
+// getSandboxesNeedingUpdating returns sandboxes where secret in status does not match secret in spec
+func (a *PasswordSecretReconciler) getSandboxesNeedingUpdating() []vapi.Sandbox {
+	sandboxes := []vapi.Sandbox{}
+	for _, sb := range a.Vdb.Spec.Sandboxes {
```

Comment:
It does not change much but I think we should use the status as it represents 100% sandboxes in the db.